### PR TITLE
Issue 1929 improve performance of token-based searches

### PR DESCRIPTION
**Pull request by punktilious**

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java [hunk 1]
```diff
@@ -1031,9 +1055,19 @@ protected SqlQueryData processInclusionCriteria(QueryParameter queryParm) throws
...
                     whereClauseSegment.append(AND).append(PARAMETER_TABLE_ALIAS).append(DOT)
-                        .append(CODE_SYSTEM_ID).append(EQ).append(codeSystemIdForResourceType);
+                        .append(CODE_SYSTEM_ID).append(EQ).append(nullCheck(codeSystemIdForResourceType));
+
+                    // For #1929 add common_token_value_id = <n> for better query optimization
+                    Long commonTokenValueId = getCommonTokenValueId(resourceTypeName, currentParmValue);
+                    if (commonTokenValueId != null) {
+                        // #1929 improves cardinality estimation
+                        // resulting in far better execution plans for many search queries. Because COMMON_TOKEN_VALUE_ID
+                        // is the primary key for the common_token_values table, we don't need the CODE_SYSTEM_ID = ? predicate.
```
Comment (michaelwschroeder) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1067]: Your comment says we don't need the CODE_SYSTEM_ID = ? predicate. Should lines 1058-1060 be moved to an else leg for the if check on line 1064?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java [hunk 2]
```diff
@@ -1112,18 +1145,29 @@ private SqlQueryData processTokenParm(QueryParameter queryParm, String tableAlia
...
                     whereClauseSegment.append(tableAlias).append(DOT).append(CODE_SYSTEM_ID).append(operator)
-                            .append(BIND_VAR);
-
-                    codeSystemId = identityCache.getCodeSystemId(value.getValueSystem());
+                            .append(nullCheck(codeSystemId));
+
+                    if (commonTokenValueId != null) {
+                        // #1929 improves cardinality estimation
+                        // resulting in far better execution plans for many search queries. Because COMMON_TOKEN_VALUE_ID
+                        // is the primary key for the common_token_values table, we don't need the CODE_SYSTEM_ID = ? predicate.
```
Comment (michaelwschroeder) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1166]: same comment as above - your comment says we don't need the CODE_SYSTEM_ID = ? predicate. Should lines 1159-1161 be moved to an else leg for the if check on line 1163?

File: fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java [hunk 3]
```diff
@@ -1,88 +0,0 @@
...
- */
-public class ReferenceParameterHandler {
-
-    //@formatter:off
-    private static final List<ParameterValueHandler> handlers = Arrays.asList(
-            new CanonicalUrlHandlerImpl(),
-            new IdHandlerImpl(),
-            new TypeIdHandlerImpl(),
-            new UrlHandlerImpl()
-        );
```
Comment (michaelwschroeder) [inline fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java:39]: Do you also want to delete ParameterValueHandler, CanonicalUrlHandlerImpl, IdHandlerImpl, TypeIdHandlerImpl, and UrlHandlerImpl since they're only used by this class?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/cache/CommonTokenValuesCacheImpl.java [hunk 4]
```diff
@@ -265,4 +265,41 @@ public void prefillCodeSystems(Map<String, Integer> codeSystems) {
...
+        // Find the code-system first
+        LinkedHashMap<String,Integer> sysMap = codeSystems.get();
+        Integer codeSystemId = sysMap != null ? sysMap.get(codeSystem) : null;
+
+        if (codeSystemId == null) {
+            // not found in the local cache, try the shared cache
+            synchronized (codeSystemsCache) {
+                codeSystemId = codeSystemsCache.get(codeSystem);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/cache/CommonTokenValuesCacheImpl.java:282]: Yep, good call.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java (hunk 1 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1067]: I meant in terms of the logical query semantics. I don't think the extra predicate will impact the cardinality estimate because the value is part of a unique index. I was hesitant to remove the predicate in case it broke something...I'll have a think. I'll either remove it or update the comment to better clarify

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java (hunk 2 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1166]: Same response as above

File: fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java (hunk 3 above)
Comment (punktilious) [reply fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java:39]: Oh, good catch. I thought they were inner classes. As standalone I'll definitely get rid.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/FHIRPersistenceJDBCCache.java [hunk 5]
```diff
@@ -6,8 +6,6 @@
```
[the remark this answers is not in the document]
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/FHIRPersistenceJDBCCache.java:6]: done

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java (hunk 1 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1072]: moved to the else as suggested

**Review by michaelwschroeder — APPROVED** [review]
LGTM - just one minor comment

File: fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java (hunk 3 above)
Comment (punktilious) [reply fhir-search/src/main/java/com/ibm/fhir/search/reference/ReferenceParameterHandler.java:39]: Removed

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java (hunk 2 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:1188]: done

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/cache/CommonTokenValuesCacheImpl.java [hunk 6]
```diff
@@ -265,4 +265,40 @@ public void prefillCodeSystems(Map<String, Integer> codeSystems) {
...
+
+        // Find the code-system first
+        Integer codeSystemId = getCodeSystemId(codeSystem);
+
+        if (codeSystemId == null) {
+            // not found in the local cache, try the shared cache
+            synchronized (codeSystemsCache) {
+                codeSystemId = codeSystemsCache.get(codeSystem);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/cache/CommonTokenValuesCacheImpl.java:281]: Agreed

File: docs/src/pages/guides/FHIRPerformanceGuide.md [hunk 7]
```diff
@@ -558,7 +558,40 @@ SELECT COUNT(R.VERSION_ID)
 
 In most cases the history queries will execute very quickly. Performance will be slower for cases where a single resource has thousands of versions. To avoid this, ingestion pipelines must ensure they only update a version when necessary.
 
-## 5.4 Search Examples
+## 5.4 Search Performance
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRPerformanceGuide.md:561]: We need to update the "index" at the top of this guide...but it was already stale before this change and so I think we could do it in a separate PR.

File: docs/src/pages/guides/FHIRPerformanceGuide.md [hunk 8]
```diff
@@ -558,7 +558,40 @@ SELECT COUNT(R.VERSION_ID)
...
+/ExplanationOfBenefit?_pretty=true&claim.priority=normal,stat,deferred&_include=ExplanationOfBenefit:claim&_include=ExplanationOfBenefit:patient&patient:Patient.birthdate=le1915
+```
+
+The above query requires a join of around 13 tables which is too many for the database to try all possible orders and so the most efficient plan is never tried. The IBM FHIR Server builds the SQL based on the order of filter predicates in the search request. This can be used along with knowledge of the data to place the most selective filter first which, in this case, is the patient `birthdate` range. Rewriting the query as follows can significantly improve the response time:
+
+```
+/ExplanationOfBenefit?_pretty=true&patient:Patient.birthdate=le1915&claim.priority=normal,stat,deferred&_include=ExplanationOfBenefit:claim&_include=ExplanationOfBenefit:patient
+```
+
+**Include Code System**
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRPerformanceGuide.md:577]: I think I added basically this same information to the **TOKEN SEARCH** part in what is now section 5.5.  We should probably merge the two.

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirResourceTableGroup.java [hunk 9]
```diff
@@ -415,15 +417,15 @@ public Table addResourceTokenRefs(List<IDatabaseObject> group, String prefix) {
 
         // logical_resources (1) ---- (*) patient_resource_token_refs (*) ---- (0|1) common_token_values
         Table tbl = Table.builder(schemaName, tableName)
-                .setVersion(FhirSchemaVersion.V0006.vid())
+                .setVersion(FhirSchemaVersion.V0008.vid())
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirResourceTableGroup.java:420]: Please try add the `schema-change` label (to either the PR or the issue or both) whenever it contains a schema change...I've found this helps when putting together the release notes.

Comment (lmsurpre) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirResourceTableGroup.java:420]: I've gone ahead and added it here.

Comment (punktilious) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/FhirResourceTableGroup.java:420]: thanks

File: docs/src/pages/guides/FHIRPerformanceGuide.md (hunk 7 above)
Comment (punktilious) [reply docs/src/pages/guides/FHIRPerformanceGuide.md:561]: ok

File: docs/src/pages/guides/FHIRPerformanceGuide.md (hunk 8 above)
Comment (punktilious) [reply docs/src/pages/guides/FHIRPerformanceGuide.md:577]: We can do this in another issue which also fixes the index from your previous comment.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 10]
```diff
@@ -1203,13 +1202,14 @@ private static void checkSearchParameterCombinations(Class<?> resourceType, List
...
             case REFERENCE: {
                 // reference
-                // [parameter]=[url]
-                // [parameter]=[url|version] - canonical url
-                // [parameter]=[type]/[id]
-                // [parameter]=[id]
+                // [parameter]=[literal] - literal reference
+                // [parameter]=[type]/[id] - relative local reference
+                // [parameter]=[base]/[type]/[id] - absolute local reference
+                // [parameter]=[id] - relativel local reference
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1208]:
```suggestion
                // [parameter]=[id] - relative local reference
```
